src/runtime_src/core/pcie/linux/shim.cpp

```diff
@@ -2211,20 +2209,23 @@ int shim::xclRegRW(bool rd, uint32_t ipIndex, uint32_t offset, uint32_t *datap)
   auto& cumap = mCuMaps[ipIndex];  // {base, size}
```

Comment:
I don't think std::map::operator[] is thread safe against insertion by other threads, so multiple threads can hit this index operator at different times while one thread is busy changing the map.

The only optimization I see we can do in this function is to move 2232-2235 into the initialization block.

Reply:
`std::vector<std::pair<uint32_t*, uint32_t>> mCuMaps;`
I think it has not to be thread safe, because mCuMaps is initalised in the constructor with 128 pairs:
`mCuMaps{128, {nullptr, 0}}`
 the vector itself is never resized only the pair values are accessed.

Reply:
Alright, yes it is a vector, but it is not thread safe for same reason.  When one thread is reading the value of mCUMaps[idx] and checking cumap.first, another thread could be in the process of populating the same pair with values.